### ISSUE TITLE
Platform dependant serialization between JS and Flash

### ISSUE THREAD

**Issue by ricklesauceur** [aaulia/msgpack-haxe#12]
Hello,

I noticed a error while trying to serialize a map between flash and JS.

The incorrect behaviour is in Encode.hx.
In the writeMap function, in the key iteration, k is encoded as a string in flash and as a int in js which led to an incorrect buffer size in output. 

Surprisingly, only the map serialization in flash works, the js buffer is not decodable.

Cheers

**Comment by ricklesauceur** [comment]
Doing a specialized function for writeMap is enough to make it works.
You also have to change the switch case above.

```	
inline function writeMapInt<V>(d:Map<Int, V>) {
		var length = 0;
		for (k in d.keys()) 
			length++;

		writeMapLength(length);
		for (k in d.keys()) { 
			encode(k);
			encode(d.get(k));
		}
	}
```

**Comment by aaulia** [comment]
So the problem here is that you want to send an IntMap back and forth between flash and JS but in flash the IntMap keys is encoded as string instead of Int. Is that it ?

A simple test case that I can run against would help. But I'll take a look at this too.

**Comment by ricklesauceur** [comment]
This is it. I can see what I can do but I am a bit overwhelmed currently. Thank you for taking a look.

**Comment by aaulia** [comment]
which haxe are you using, 3.1.3. or 3.2 ?

**Comment by ricklesauceur** [comment]
I am on 3.2.

**Comment by aaulia** [comment]
Hmm I've tried encoding IntMap<String> on Flash, and the result is that the keys are correctly encoded as TInt and not as String. I've also verify the byte output.

**Comment by ricklesauceur** [comment]
I try both using IntMap and Map. The result was the same. 
If you check the length it should not be the same because in one case, key are encoded as string, and in the other side, keys are encoded as int.

With my change it works. I do not really know what to add.

**Comment by aaulia** [comment]
Can you give me a dummy data structure based on what you want to serialize? I've tried with an IntMap with String value and the length on both Flash and JS and its content is identical.

    var data = new IntMap<String>();
    data.set(10, "ten");
    data.set(11, "eleven");
    data.set(12, "twelve");
    
    var out = MsgPack.encode(data);
    trace(out.toHex());

This produce identical result both in Flash and JS.

Oh and one more thing, did you use the DecodeOption when decoding? In case om Map, you can decode it as an Object or as a Map (by default it decode to Object).

**Comment by ricklesauceur** [comment]
I decode with the map option.

The data structure is as follow:

var message = { OwnerID:component.ownerID, ID:component.ID, ClassID:component.classID, object:buffer };
var message_buffer = MsgPack.encode(message);

The buffer object in this case is (all values are float):

		var input = {
			pos_x : pos.x,
			pos_y : pos.y,
			pos_z : pos.z,
			rot_x : rot.x,
			rot_y : rot.y,
			rot_z : rot.z,
			rot_w : rot.w
		};
		
		return MsgPack.encode(input);

**Comment by aaulia** [comment]
Hmm, at a glance, all this is an Anonymous Structure (Object) and no a Map. All the key should be encoded to String, no? I'll try it out and see the result. My guess is if you decode it with the map option you should get a StringMap<Dynamic> IIRC.

So just to clarify,

    component.ownerID
    component.ID
    component.classID

is an Int or Int64, while

    buffer

is a Bytes from encoding the ``input``. Right?

**Comment by ricklesauceur** [comment]
I have forgot the interesting part (i am a bit tired sorry).

The variable input is encoded fine. The part which is not is in between:

		var componentMap = new Map<Int, haxe.io.Bytes>();
		
		for (component in networkComponent.networkComponents)
		{
			var buffer = component.serialize();
			componentMap.set(component.GetClassID(), buffer);
		}
		
		var map_buffer = MsgPack.encode(componentMap);
		return map_buffer;

So to recap, I remember that after debugging this map was the problem.

The buffer returned by  component.serialize(); is the anonymous structure with position/rotation.

**Comment by aaulia** [comment]
Okay, I've tried mocking the data using your description. It seems that Flash and JS is producing different bytes. Thanks for this. I will investigate further.

**Comment by aaulia** [comment]
I've completed my investigation, I've logged every step in the encoding process. Unfortunately I have not found the bug you described.

When encoding the same data, Flash output is indeed different from the JS output. But the difference is caused by how Flash treated the Anonymous Structure. 

So for example, in Flash this is the order which the ''input'' is encoded

    pos_x
    rot_x
    pos_y
    rot_y
    rot_z
    pos_z
    rot_w

While in JS the order is

    pos_x
    pos_y
    pos_z
    rot_x
    rot_y
    rot_z
    rot_w

this caused the resulting Bytes to differ (because the string key order is different). But the content is still the same.

Both Flash and JS encode the map correctly (Int Key and Bytes Value).

**Comment by ricklesauceur** [comment]
The order is not important and should not change the final size of the buffer. 

The problem is the map with component:

    var componentMap = new Map<Int, haxe.io.Bytes>();

    for (component in networkComponent.networkComponents)
    {
        var buffer = component.serialize();
        componentMap.set(component.GetClassID(), buffer);
    }

    var map_buffer = MsgPack.encode(componentMap);
    return map_buffer;

The Int key is going to get encoded as a String on flash I think and an Int in Javascript.

**Comment by aaulia** [comment]
So far my test shows that both Flash and JS encode the key correctly as Int. Here is what I use to test it (based on your description of the problem and code).

`build.hxml` I copied the msgpack-haxe source folder into the project for logging, so ignore the commented out `-lib` line.
````
-cp src 
-main Main
-swf bin/swf/Main.swf
-swf-header 400:300:60:FFFFFF
-debug 
# -lib msgpack-haxe

--next

-cp src 
-main Main
-js bin/js/Main.js
-debug 
# -lib msgpack-haxe
````

`Main.hx`
````
package;

import haxe.io.Bytes;
import org.msgpack.MsgPack;

class Position {
    public var x:Float;
    public var y:Float;
    public var z:Float;

    public function new(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
}

class Rotation {
    public var x:Float;
    public var y:Float;
    public var z:Float;
    public var w:Float;

    public function new(x, y, z, w) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.w = w;
    }
}

class Component {
    private var classID:Int;
    private var pos:Position;
    private var rot:Rotation;

    public function getClassID() return classID;

    public function new(classID, pos, rot) {
        this.classID = classID;
        this.pos     = pos;
        this.rot     = rot;
    }

    public function serialize() {
        return MsgPack.encode({
                pos_x: pos.x,
                pos_y: pos.y,
                pos_z: pos.z,
                
                rot_x: rot.x,
                rot_y: rot.y,
                rot_z: rot.z,
                rot_w: rot.w
            });
    }
}

class Main {
    public static function main() {
        var networkComponents = [
            new Component(0, new Position(1, 2, 3), new Rotation(90,  0,  0, 1)),
            new Component(1, new Position(4, 5, 6), new Rotation( 0, 90,  0, 1)),
            new Component(2, new Position(7, 8, 9), new Rotation( 0,  0, 90, 1))
        ];

        var componentMap = new Map<Int, Bytes>();
        for (component in networkComponents) {
            var buffer = component.serialize();
            componentMap.set(component.getClassID(), buffer);
        }

        var mapBuffer = MsgPack.encode(componentMap);
        Console.log('${mapBuffer.length}\n${mapBuffer.toHex()}');
    }
}
````

Here is the result (with all the encoding log) from Flash (I dumped it to Chrome console for better logging).

````
Encoding Object => 7
    Encoding String => "pos_y"
    Encoding Int => 2
    Encoding String => "rot_w"
    Encoding Int => 1
    Encoding String => "rot_z"
    Encoding Int => 0
    Encoding String => "rot_y"
    Encoding Int => 0
    Encoding String => "pos_z"
    Encoding Int => 3
    Encoding String => "pos_x"
    Encoding Int => 1
    Encoding String => "rot_x"
    Encoding Int => 90
Encoding Object => 7
    Encoding String => "pos_y"
    Encoding Int => 5
    Encoding String => "rot_w"
    Encoding Int => 1
    Encoding String => "rot_z"
    Encoding Int => 0
    Encoding String => "rot_y"
    Encoding Int => 90
    Encoding String => "pos_z"
    Encoding Int => 6
    Encoding String => "pos_x"
    Encoding Int => 4
    Encoding String => "rot_x"
    Encoding Int => 0
Encoding Object => 7
    Encoding String => "pos_y"
    Encoding Int => 8
    Encoding String => "rot_w"
    Encoding Int => 1
    Encoding String => "rot_z"
    Encoding Int => 90
    Encoding String => "rot_y"
    Encoding Int => 0
    Encoding String => "pos_z"
    Encoding Int => 9
    Encoding String => "pos_x"
    Encoding Int => 7
    Encoding String => "rot_x"
    Encoding Int => 0
Encoding Map => 3
    Encoding Int => 0
    Encoding Bytes => {50 => 87a5706f735f7902a5726f745f7701a5726f745f7a00a5726f745f7900a5706f735f7a03a5706f735f7801a5726f745f785a}
    Encoding Int => 1
    Encoding Bytes => {50 => 87a5706f735f7905a5726f745f7701a5726f745f7a00a5726f745f795aa5706f735f7a06a5706f735f7804a5726f745f7800}
    Encoding Int => 2
    Encoding Bytes => {50 => 87a5706f735f7908a5726f745f7701a5726f745f7a5aa5726f745f7900a5706f735f7a09a5706f735f7807a5726f745f7800}
160
8300c43287a5706f735f7902a5726f745f7701a5726f745f7a00a5726f745f7900a5706f735f7a03a5706f735f7801a5726f745f785a01c43287a5706f735f7905a5726f745f7701a5726f745f7a00a5726f745f795aa5706f735f7a06a5706f735f7804a5726f745f780002c43287a5706f735f7908a5726f745f7701a5726f745f7a5aa5726f745f7900a5706f735f7a09a5706f735f7807a5726f745f7800
````

and here is the JS log

````
Encoding Object => 7
    Encoding String => "pos_x"
    Encoding Int => 1
    Encoding String => "pos_y"
    Encoding Int => 2
    Encoding String => "pos_z"
    Encoding Int => 3
    Encoding String => "rot_x"
    Encoding Int => 90
    Encoding String => "rot_y"
    Encoding Int => 0
    Encoding String => "rot_z"
    Encoding Int => 0
    Encoding String => "rot_w"
    Encoding Int => 1
Encoding Object => 7
    Encoding String => "pos_x"
    Encoding Int => 4
    Encoding String => "pos_y"
    Encoding Int => 5
    Encoding String => "pos_z"
    Encoding Int => 6
    Encoding String => "rot_x"
    Encoding Int => 0
    Encoding String => "rot_y"
    Encoding Int => 90
    Encoding String => "rot_z"
    Encoding Int => 0
    Encoding String => "rot_w"
    Encoding Int => 1
Encoding Object => 7
    Encoding String => "pos_x"
    Encoding Int => 7
    Encoding String => "pos_y"
    Encoding Int => 8
    Encoding String => "pos_z"
    Encoding Int => 9
    Encoding String => "rot_x"
    Encoding Int => 0
    Encoding String => "rot_y"
    Encoding Int => 0
    Encoding String => "rot_z"
    Encoding Int => 90
    Encoding String => "rot_w"
    Encoding Int => 1
Encoding Map => 3
    Encoding Int => 0
    Encoding Bytes => {50 => 87a5706f735f7801a5706f735f7902a5706f735f7a03a5726f745f785aa5726f745f7900a5726f745f7a00a5726f745f7701}
    Encoding Int => 1
    Encoding Bytes => {50 => 87a5706f735f7804a5706f735f7905a5706f735f7a06a5726f745f7800a5726f745f795aa5726f745f7a00a5726f745f7701}
    Encoding Int => 2
    Encoding Bytes => {50 => 87a5706f735f7807a5706f735f7908a5706f735f7a09a5726f745f7800a5726f745f7900a5726f745f7a5aa5726f745f7701}
160
8300c43287a5706f735f7801a5706f735f7902a5706f735f7a03a5726f745f785aa5726f745f7900a5726f745f7a00a5726f745f770101c43287a5706f735f7804a5706f735f7905a5706f735f7a06a5726f745f7800a5726f745f795aa5726f745f7a00a5726f745f770102c43287a5706f735f7807a5706f735f7908a5706f735f7a09a5726f745f7800a5726f745f7900a5726f745f7a5aa5726f745f7701
````

As you can see, on both target, the Bytes length is the same, the order is different but overall both encode the key and value correctly. Without minimal reproducible sample, I don't know how else am I going to proceed with this.

**Comment by ricklesauceur** [comment]
Let me try a repro and I come back to you.
